src/node/eviction.cpp

```diff
@@ -172,39 +183,56 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
+[[nodiscard]] std::optional<NodeId> SelectNodeToEvict(std::vector<NodeEvictionCandidate>&& vEvictionCandidates, bool force)
 {
     // Protect connections with certain characteristics
 
     ProtectNoBanConnections(vEvictionCandidates);
 
     ProtectOutboundConnections(vEvictionCandidates);
 
+    // Keep track of the last protected candidate so we can evict it if forced
+    NodeEvictionCandidate last_out;
+
     // Deterministically select 4 peers to protect by netgroup.
     // An attacker cannot predict which netgroups will be protected
-    EraseLastKElements(vEvictionCandidates, CompareNetGroupKeyed, 4);
+    auto last = EraseLastKElements(vEvictionCandidates, CompareNetGroupKeyed, 4);
+    if (last.has_value()) last_out = last.value();
     // Protect the 8 nodes with the lowest minimum ping time.
     // An attacker cannot manipulate this metric without physically moving nodes closer to the target.
-    EraseLastKElements(vEvictionCandidates, ReverseCompareNodeMinPingTime, 8);
+    last = EraseLastKElements(vEvictionCandidates, ReverseCompareNodeMinPingTime, 8);
+    if (last.has_value()) last_out = last.value();
     // Protect 4 nodes that most recently sent us novel transactions accepted into our mempool.
     // An attacker cannot manipulate this metric without performing useful work.
-    EraseLastKElements(vEvictionCandidates, CompareNodeTXTime, 4);
+    last = EraseLastKElements(vEvictionCandidates, CompareNodeTXTime, 4);
+    if (last.has_value()) last_out = last.value();
     // Protect up to 8 non-tx-relay peers that have sent us novel blocks.
-    EraseLastKElements(vEvictionCandidates, CompareNodeBlockRelayOnlyTime, 8,
-                       [](const NodeEvictionCandidate& n) { return !n.m_relay_txs && n.fRelevantServices; });
+    last = EraseLastKElements(vEvictionCandidates, CompareNodeBlockRelayOnlyTime, 8,
+                              [](const NodeEvictionCandidate& n) { return !n.m_relay_txs && n.fRelevantServices; });
+    if (last.has_value()) last_out = last.value();
 
     // Protect 4 nodes that most recently sent us novel blocks.
     // An attacker cannot manipulate this metric without performing useful work.
-    EraseLastKElements(vEvictionCandidates, CompareNodeBlockTime, 4);
+    last = EraseLastKElements(vEvictionCandidates, CompareNodeBlockTime, 4);
+    if (last.has_value()) last_out = last.value();
 
     // Protect some of the remaining eviction candidates by ratios of desirable
     // or disadvantaged characteristics.
-    ProtectEvictionCandidatesByRatio(vEvictionCandidates);
-
-    if (vEvictionCandidates.empty()) return std::nullopt;
+    last = ProtectEvictionCandidatesByRatio(vEvictionCandidates);
+    if (last.has_value()) last_out = last.value();
+
+    if (vEvictionCandidates.empty()) {
+        if (force) {
+            return last_out.id;
+        } else {
+            return std::nullopt;
+        }
```

Comment:
nit
```suggestion
        return force ? std::make_optional(last_out.id) : std::nullopt;
```

Reply:
Not really familiar with this stuff, but that seems to be an MSVC bug? https://stackoverflow.com/questions/55152552/visual-studio-not-performing-rvo-when-ternary-operator-is-used-and-move-copy-cto

We do ternary operators in plenty of places, don't think we need to workaround this?

Reply:
Does RVO even apply here? I thought it applies to complex structures, a large vector being the canonical example, where we don't want to copy every element. But here we're just returning a `NodeId` which is a `int64_t`, wrapped by `std::optional`. I don't think there's any savings by moving instead of copying one of these.